src/execution_tree/primitives/argmax.cpp

```diff
@@ -0,0 +1,323 @@
+            }
+
+            ///////////////////////////////////////////////////////////////////////////
+            primitive_argument_type argmax1d(args_type && args) const
+            {
+                // `axis` is optional
+                if (args.size() == 2)
+                {
+                    // `axis` must be a scalar if provided
+                    if (args[1].num_dimensions() != 0)
+                    {
+                        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                            "argmax::argmax2d",
+                            generate_error_message(
+                                "operand axis must be a scalar", name_,
+                                codename_));
+                    }
+                    const int axis = args[1].scalar();
+                    // `axis` can only be -1 or 0
+                    if (axis < -1 || axis > 0)
+                    {
+                        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                            "argmax::argmax2d",
+                            generate_error_message(
+                                "operand axis can only between -1 and 0 for "
+                                "an a operand that is 1d",
+                                name_, codename_));
+                    }
+                }
+
+                auto a = args[0].vector();
+
+                if (a.size() == 0)
+                {
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "argmax::argmax2d",
+                        generate_error_message(
+                            "attempt to get argmax of an empty sequence",
+                            name_, codename_));
+                }
+
+                const auto max_it = std::max_element(a.begin(), a.end());
+
+                return std::distance(a.begin(), max_it);
+            }
+
+            ///////////////////////////////////////////////////////////////////////////
+            primitive_argument_type argmax2d_flatten(arg_type && arg_a) const
+            {
+                using phylanx::util::matrix_row_iterator;
+
+                auto a = arg_a.matrix();
+
+                if (a.rows() == 0)
+                {
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "argmax::argmax2d",
+                        generate_error_message(
+                            "attempt to get argmax of an empty sequence",
+                            name_, codename_));
+                }
+
+                const matrix_row_iterator<decltype(a)> a_begin(a);
+                const matrix_row_iterator<decltype(a)> a_end(a, a.rows());
+
+                double global_max = 0.;
+                std::size_t global_index = 0ul;
+                std::size_t passed_rows = 0ul;
+                for (auto it = a_begin; it != a_end; ++it, ++passed_rows)
+                {
+                    const auto local_max = std::max_element(it->begin(), it->end());
```

Comment:
Could it happen that `it->begin() == it->end()`? In this case the returned iterator shouldn't be dereferenced.

Reply:
No. Because the row count is checked in line 129-136

Reply:
That check tests for whether there are rows, while I was asking about the number of elements in  row.

Reply:
I don't know if that can happen. I added a bit that checks the column count for good measure.